DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/optimize_inner_loop/src/optimize_inner_loop.cpp

```diff
@@ -0,0 +1,280 @@
+//==============================================================
+// Copyright Intel Corporation
+//
+// SPDX-License-Identifier: MIT
+// =============================================================
+#include <array>
+#include <iomanip>
+#include <iostream>
+#include <random>
+#include <type_traits>
+
+#include <CL/sycl.hpp>
+#include <CL/sycl/INTEL/fpga_extensions.hpp>
+
+// dpc_common.hpp can be found in the dev-utilities include folder.
+// e.g., $ONEAPI_ROOT/dev-utilities//include/dpc_common.hpp
+#include "dpc_common.hpp"
+
+using namespace sycl;
+
+// allow the maximum random number value to be controlled from the command line
+#ifndef RAND_RANGE_MAX
+#define RAND_RANGE_MAX 3
+#endif
+
+//// constants
+constexpr int kNumKernels = 3;
+constexpr int kRandRangeMax = RAND_RANGE_MAX;
+constexpr double kProbSuccess = 1.0 / (double)(kRandRangeMax);
+
+// Declare the kernel class names globally to avoid name mangling.
+// Templating allows us to instantiate multiple versions of the kernel.
+template <int Version> class Producer;
+template <int Version> class Consumer;
+
+// Declare the pipe class name globally to avoid name mangling.
+// Templating allows us to instantiate multiple versions of pipes for each 
+// version of the kernel.
+template <int Version> class PipeClass;
+
+//
+// Submits the kernel, which is templated on the variables:
+//    Version             - The version ID of the kernel
+//    InElementUpperBound - The upperbound (inclusive) on the elements of the
+//                          'in' vector (a negative value implies no bound). In
+//                          other words: if InElementUpperBound >= 0, then
+//                          in[i] <= InElementUpperBound, for all elements
+//                          of 'in'
+//    SpecIters           - The number of speculated iterations to set for the
+//                          inner loop
+//
+template <int Version, int InElementUpperBound, int SpecIters>
```

Comment:
Naming should be template\<int version int in_element_upper_bound, int spec_iters\>